sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala

```diff
@@ -1407,13 +1409,62 @@ private[hive] object HiveClientImpl extends Logging {
       case _ =>
         new HiveConf(conf, classOf[HiveConf])
     }
-    try {
+    val hive = try {
       Hive.getWithoutRegisterFns(hiveConf)
     } catch {
       // SPARK-37069: not all Hive versions have the above method (e.g., Hive 2.3.9 has it but
-      // 2.3.8 don't), therefore here we fallback when encountering the exception.
+      // 2.3.8 doesn't), therefore here we fallback when encountering the exception.
       case _: NoSuchMethodError =>
         Hive.get(hiveConf)
     }
+    configureMaxThriftMessageSize(hiveConf, hive.getMSC)
+    hive
+  }
+
+  // SPARK-49489: a surgery for Hive 2.3.10 due to lack of HIVE-26633
+  private def configureMaxThriftMessageSize(
+      hiveConf: HiveConf, msClient: IMetaStoreClient): Unit = try {
+    msClient match {
+      // Hive uses Java Dynamic Proxy to enhance the MetaStoreClient to support synchronization
+      // and retrying, we should unwrap and access the real MetaStoreClient instance firstly
+      case proxy if JdkProxy.isProxyClass(proxy.getClass) =>
+        JdkProxy.getInvocationHandler(proxy) match {
+          case syncHandler if syncHandler.getClass.getName.endsWith("SynchronizedHandler") =>
+            val realMscField = SparkClassUtils.classForName(
```

Comment:
Perhaps we can directly use `syncHandler.getClass().getDeclaredField("client")` here, and there should be no practical difference in effect.

Reply:
Then, perhaps we can try to refactor a bit here,  maybe like

```scala
JdkProxy.getInvocationHandler(proxy) match {
    case syncHandler if syncHandler.getClass.getName.endsWith("SynchronizedHandler") =>
      val realMsc = getFieldValue(syncHandler, "client").asInstanceOf[IMetaStoreClient]
      configureMaxThriftMessageSize(hiveConf, realMsc)

    case retryHandler: RetryingMetaStoreClient =>
      val realMsc = getFieldValue(retryHandler, "base").asInstanceOf[IMetaStoreClient]
      configureMaxThriftMessageSize(hiveConf, realMsc)

    case _ => // do nothing
  }

private def getFieldValue(obj: AnyRef, fieldName: String): AnyRef = {
  val field = obj.getClass.getDeclaredField(fieldName)
  field.setAccessible(true)
  field.get(obj)
}
```

Reply:
@LuciferYang thanks for the review, I simplify the reflection call by following your suggestions.

For concerns about the reflection code itself, I tune the code to make the added code only takes effect when the user configures `hive.thrift.client.max.message.size` explicitly, in case users compile the Spark with their own modified Hive version.